File: apps/web/src/lib/permissions.ts
export const permissions = [
  "DASHBOARD_READ",
  "AUDIT_LOG_READ",
  "NETWORK_MAP_READ",
  "USER_READ",
  "USER_CREATE",
  "USER_UPDATE",
  "USER_DISABLE",
  "USER_ENABLE",
  "USER_PASSWORD_RESET",
  "SETTINGS_READ",
  "SETTINGS_UPDATE",
  "SERVER_READ",
  "SERVER_CREATE",
  "SERVER_UPDATE",
  "SERVER_DELETE",
  "SERVICE_READ",
  "SERVICE_CREATE",
  "SERVICE_UPDATE",
  "SERVICE_DELETE",
  "AGENT_READ",
  "AGENT_ENROLL",
  "AGENT_CANCEL_ENROLLMENT",
  "AGENT_ROTATE_TOKEN",
  "AGENT_DECOMMISSION",
] as const;

export type UserPermission = (typeof permissions)[number];

export type PermissionPreset = "ADMIN" | "OPERATOR" | "VIEWER";

export const permissionPresets: Record<PermissionPreset, UserPermission[]> = {
  ADMIN: [...permissions],
  OPERATOR: [
    "DASHBOARD_READ",
    "AUDIT_LOG_READ",
    "NETWORK_MAP_READ",
    "SERVER_READ",
    "SERVER_CREATE",
    "SERVER_UPDATE",
    "SERVICE_READ",
    "SERVICE_CREATE",
    "SERVICE_UPDATE",
    "SERVICE_DELETE",
    "AGENT_READ",
    "AGENT_ENROLL",
    "AGENT_CANCEL_ENROLLMENT",
    "AGENT_ROTATE_TOKEN",
    "AGENT_DECOMMISSION",
  ],
  VIEWER: [
    "DASHBOARD_READ",
    "AUDIT_LOG_READ",
    "NETWORK_MAP_READ",
    "SERVER_READ",
    "SERVICE_READ",
    "AGENT_READ",
  ],
};

export function isUserPermission(
  permission: string | null | undefined
): permission is UserPermission {
  return permissions.includes(permission as UserPermission);
}

export function hasPermission(
  assigned: readonly string[] | null | undefined,
  permission: UserPermission
) {
  return Array.isArray(assigned) && assigned.includes(permission);
}

export function can(
  assigned: readonly string[] | null | undefined,
  permission: UserPermission
) {
  return hasPermission(assigned, permission);
}

export function canAccessUserManagement(
  assigned: readonly string[] | null | undefined
) {
  return (
    can(assigned, "USER_READ") ||
    can(assigned, "USER_CREATE") ||
    can(assigned, "USER_UPDATE") ||
    can(assigned, "USER_DISABLE") ||
    can(assigned, "USER_ENABLE") ||
    can(assigned, "USER_PASSWORD_RESET")
  );
}

export function canReadDashboard(assigned: readonly string[] | null | undefined) {
  return can(assigned, "DASHBOARD_READ");
}

export function canReadAuditLogs(assigned: readonly string[] | null | undefined) {
  return can(assigned, "AUDIT_LOG_READ");
}

export function canReadNetworkMap(assigned: readonly string[] | null | undefined) {
  return can(assigned, "NETWORK_MAP_READ");
}

export function canReadSettings(assigned: readonly string[] | null | undefined) {
  return can(assigned, "SETTINGS_READ");
}

export function canUpdateSettings(assigned: readonly string[] | null | undefined) {
  return can(assigned, "SETTINGS_UPDATE");
}

export function canReadServers(assigned: readonly string[] | null | undefined) {
  return can(assigned, "SERVER_READ");
}

export function canCreateServers(assigned: readonly string[] | null | undefined) {
  return can(assigned, "SERVER_CREATE");
}

export function canUpdateServers(assigned: readonly string[] | null | undefined) {
  return can(assigned, "SERVER_UPDATE");
}

export function canDeleteServers(assigned: readonly string[] | null | undefined) {
  return can(assigned, "SERVER_DELETE");
}

export function canReadServices(assigned: readonly string[] | null | undefined) {
  return can(assigned, "SERVICE_READ");
}

export function canCreateServices(assigned: readonly string[] | null | undefined) {
  return can(assigned, "SERVICE_CREATE");
}

export function canUpdateServices(assigned: readonly string[] | null | undefined) {
  return can(assigned, "SERVICE_UPDATE");
}

export function canDeleteServices(assigned: readonly string[] | null | undefined) {
  return can(assigned, "SERVICE_DELETE");
}

export function canReadAgents(assigned: readonly string[] | null | undefined) {
  return can(assigned, "AGENT_READ");
}

export function canEnrollAgents(assigned: readonly string[] | null | undefined) {
  return can(assigned, "AGENT_ENROLL");
}

export function canCancelAgentEnrollments(
  assigned: readonly string[] | null | undefined
) {
  return can(assigned, "AGENT_CANCEL_ENROLLMENT");
}

export function canRotateAgentTokens(
  assigned: readonly string[] | null | undefined
) {
  return can(assigned, "AGENT_ROTATE_TOKEN");
}

export function canDecommissionAgents(
  assigned: readonly string[] | null | undefined
) {
  return can(assigned, "AGENT_DECOMMISSION");
}

export function canViewOnly(assigned: readonly string[] | null | undefined) {
  return (
    Array.isArray(assigned) &&
    assigned.every((permission) =>
      [
        "DASHBOARD_READ",
        "AUDIT_LOG_READ",
        "NETWORK_MAP_READ",
        "SERVER_READ",
        "SERVICE_READ",
        "AGENT_READ",
      ].includes(permission)
    )
  );
}

File: apps/api/src/main/java/dev/kyvora/api/config/security/Permissions.java
package dev.kyvora.api.config.security;

import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Component;

import dev.kyvora.api.auth.entity.UserPermission;

@Component("permissions")
public class Permissions {

	public boolean canReadDashboard(Authentication authentication) {
		return hasPermission(authentication, UserPermission.DASHBOARD_READ);
	}

	public boolean canReadAuditLogs(Authentication authentication) {
		return hasPermission(authentication, UserPermission.AUDIT_LOG_READ);
	}

	public boolean canReadNetworkMap(Authentication authentication) {
		return hasPermission(authentication, UserPermission.NETWORK_MAP_READ);
	}

	public boolean canReadUsers(Authentication authentication) {
		return hasPermission(authentication, UserPermission.USER_READ);
	}

	public boolean canCreateUsers(Authentication authentication) {
		return hasPermission(authentication, UserPermission.USER_CREATE);
	}

	public boolean canUpdateUsers(Authentication authentication) {
		return hasPermission(authentication, UserPermission.USER_UPDATE);
	}

	public boolean canDisableUsers(Authentication authentication) {
		return hasPermission(authentication, UserPermission.USER_DISABLE);
	}

	public boolean canEnableUsers(Authentication authentication) {
		return hasPermission(authentication, UserPermission.USER_ENABLE);
	}

	public boolean canResetUserPasswords(Authentication authentication) {
		return hasPermission(authentication, UserPermission.USER_PASSWORD_RESET);
	}

	public boolean canAccessUserManagement(Authentication authentication) {
		return hasAnyPermission(
				authentication,
				UserPermission.USER_READ,
				UserPermission.USER_CREATE,
				UserPermission.USER_UPDATE,
				UserPermission.USER_DISABLE,
				UserPermission.USER_ENABLE,
				UserPermission.USER_PASSWORD_RESET);
	}

	public boolean canReadSettings(Authentication authentication) {
		return hasPermission(authentication, UserPermission.SETTINGS_READ);
	}

	public boolean canUpdateSettings(Authentication authentication) {
		return hasPermission(authentication, UserPermission.SETTINGS_UPDATE);
	}

	public boolean canReadServers(Authentication authentication) {
		return hasPermission(authentication, UserPermission.SERVER_READ);
	}

	public boolean canCreateServers(Authentication authentication) {
		return hasPermission(authentication, UserPermission.SERVER_CREATE);
	}

	public boolean canUpdateServers(Authentication authentication) {
		return hasPermission(authentication, UserPermission.SERVER_UPDATE);
	}

	public boolean canDeleteServers(Authentication authentication) {
		return hasPermission(authentication, UserPermission.SERVER_DELETE);
	}

	public boolean canReadServices(Authentication authentication) {
		return hasPermission(authentication, UserPermission.SERVICE_READ);
	}

	public boolean canCreateServices(Authentication authentication) {
		return hasPermission(authentication, UserPermission.SERVICE_CREATE);
	}

	public boolean canUpdateServices(Authentication authentication) {
		return hasPermission(authentication, UserPermission.SERVICE_UPDATE);
	}

	public boolean canDeleteServices(Authentication authentication) {
		return hasPermission(authentication, UserPermission.SERVICE_DELETE);
	}

	public boolean canReadAgents(Authentication authentication) {
		return hasPermission(authentication, UserPermission.AGENT_READ);
	}

	public boolean canEnrollAgents(Authentication authentication) {
		return hasPermission(authentication, UserPermission.AGENT_ENROLL);
	}

	public boolean canPullAgents(Authentication authentication) {
		return hasPermission(authentication, UserPermission.AGENT_PULL);
	}

	public boolean canDecommissionAgents(Authentication authentication) {
		return hasPermission(authentication, UserPermission.AGENT_DECOMMISSION);
	}

	private boolean hasPermission(Authentication authentication, UserPermission permission) {
		if (authentication == null || !authentication.isAuthenticated()) {
			return false;
		}
		String authority = permission.authority();
		return authentication.getAuthorities().stream()
				.anyMatch(grantedAuthority -> authority.equals(grantedAuthority.getAuthority()));
	}

	private boolean hasAnyPermission(Authentication authentication, UserPermission... permissions) {
		for (UserPermission permission : permissions) {
			if (hasPermission(authentication, permission)) {
				return true;
			}
		}
		return false;
	}
}
